prime-router/docs/proposals/0003-lineage/additional-notes.md

```diff
@@ -23,17 +23,40 @@ One option would be to implement triggers that prevent updates except for a stat
+
+**ReportFunction and BatchFunction**:  As they do transforms on the data, both of these functions have access to both parent and child Report.kt objects.  These functions uses that Report.kt data to create item_lineage info.  Notes that Reports can be transmogrified in memory in all kinds of ways in these steps, through split(), merge(), copy(), and filter().   Item_lineage is carefully preserved through all such transformations.   If in the future we add new clever data transform code to Report.kt, we will need to make sure item_lineage is preserved during that transform.
+
+**SendFunction and DownloadFunction**:  These functions never create Report objects (which is good - we should be messing with the data as little as possible).  In these cases, item_lineage data for the parent report is read from the database, and that's used as the basis for creating the parent->child item_lineage.   Currently, these functions are also simpler, in that there's a 1:1 mapping from parent to child rows in these functions.
+
+**Splitting and merging individual items**:  While reports break apart and merge, etc, items do not.  That is, as of this writing, we have no use cases where *rows* split or merge.  While the database is designed to handle these situations, the code has not been tested for this, and I suspect there will be issues.
+
+**RedoxTransport special case**: This is currently the only Transport that can fail/succeed on individual items.  The item_lineage.transport_result column is designed to store these fine-grained send results.   The retry mechanism adds complexity here as well.   Each time a Redox report is sent, its rows can have one of three possible statuses:
+1. Success
+2. Failed to send
+3. Not sent, because it was already successfully sent in some previous attempt.
+
+Note that because of the 'all or nothing' assertion above, the retry code in RedoxTransport was modified to track item_lineage for all three of these statuses.
+
+**Main issue remaining**:  When a Report is read into BatchFunction whose schema has no trackingElement defined in the .schema, we lose the tracking_id column in the item_lineage table.  This is an artifact of how lineage tracking works in BatchFunction, where lineage is created in-memory based on data in the Report.kt objects.   If the parent report in the lineage has no trackingElement defined, then of course we can't fill in the item_lineage.tracking_id column.  (Currently most/all redox and hl7 schemas are affected by this, as they do not have trackingElements.)  A possible fix is to read in the parent's item_lineage from the database table.  However, that ID often does not exist in the child data - its not clear what the value is of tracking an ID that is not even in the data.  Further complexity with trackingElements will arise if we begin splitting/merging Items (see comment above).
+
+
```

Comment:
Perhaps we should require a tracking element in all schemas? If one is not valid, we should generate one?